Fix several bugs: in raft, in qsync, in box.ctl.promote()

Gerold103 opened the pull request:
The patchset fixes several unrelated bugs. 2 of them are from the ticket #7253. The ticket contains 3 issues whose solution was designed in an RFC, but one of the bugs couldn't be fixed like proposed. Explanation why and possible solutions will come soon in that RFC and maybe in the ticket.

At the moment of submission I get "error 500" when open any CI job and "debug" just hangs in yellow state. Sending it for review now hoping the debug CI passes later.

Part-of #7253

sergepetrenko commented on test/replication-luatest/gh_7253_election_long_wal_write_test.lua:
```diff
@@ -0,0 +1,173 @@
+            server.build_instance_uri('server3'),
+        },
+    }
+    box_cfg.listen = box_cfg.replication[1]
```
You don't have to set `listen` manually, it's already in the default server configuration (`test/instances/default.lua`)

Gerold103 replied:
Thanks, done.

sergepetrenko commented on test/replication-luatest/gh_7253_election_long_wal_write_test.lua:
```diff
@@ -0,0 +1,173 @@
+    end))
+    --
+    -- Cleanup.
+    --
```
Let's maybe put it into `g.after_test` ?
Looks like a good example where g.after_test is actually useful.

Gerold103 replied:
I don't really like splitting the body of a test case into multiple functions. I wouldn't mind if this would be not related to the test, but in the end I have `wait_fullmesh(g)` which should fail if split-brain happened and this is a part of the test body.

On a broader topic, I don't see any sense in `after_test` for cleaning up something test-specific at all. I suppose it could try to restart instances or restore config if there are many test cases and we really-really want as many of them to run as possible even if one failed. But otherwise it is just a test, simply let it fail. It would work just fine if luatest wouldn't loose the entire .result file if the test hangs even on the last line. I've lost count how many hours I lost on it. As well as on tests having `after_test` and hanging somewhere - they won't produce any .result file either.

sergepetrenko replied:
> It would work just fine if luatest wouldn't loose the entire .result file if the test hangs even on the last line. I've lost count how many hours I lost on it. As well as on tests having after_test and hanging somewhere - they won't produce any .result file either.

I didn't know that `after_test` was the culprit.

Anyway, IMO having prepare/teardown code in before_test/after_test looks more clean. Leave it as is if you like.

Gerold103 replied:
It is not the culprit if you mean not producing any .result file. Not having `after_test` won't produce a .result file on hang either. The problem is that `after_test` doesn't help to fix it. Thus I simply don't see any sense in it except for doing common cleanup for many test cases at once.

sergepetrenko replied:
>  Not having after_test won't produce a .result file on hang either.

Ok, I see now.

sergepetrenko commented on test/replication-luatest/gh_7253_election_long_wal_write_test.lua:
```diff
@@ -0,0 +1,173 @@
+    -- Server1 receives an ack and blocks on CONFIRM WAL write.
+    --
+    server_block_next_wal_write(g.server1)
+    server_unblock_wal_write(g.server2)
```
It seems you could test this with only two servers:
servers 1 and 2 would act the same, and server 2 would bump the term instead of server3.

Gerold103 replied:
Indeed, thanks for the comment, done.

sergepetrenko commented on src/lib/raft/raft.c:
```diff
@@ -158,6 +159,32 @@ raft_add_vote(struct raft *raft, int src, int dst)
+ * to any other instance yet.
+ */
+static void
+raft_del_vote(struct raft *raft, int src, int dst)
```
Maybe `raft_revoke_vote` or `raft_vote_revoke`?

Gerold103 replied:
I used `raft_voke_vote`. It is consistent with other `raft_<verb>vote` functions like `raft_add_vote`, `raft_sm_schedule_new_vote`, `raft_sm_try_new_vote`.

sergepetrenko commented on src/lib/raft/raft.c:
```diff
@@ -158,6 +159,32 @@ raft_add_vote(struct raft *raft, int src, int dst)
+	bool was_max = v->count == raft->max_vote;
+	--v->count;
+	if (was_max) {
+		raft->max_vote = 0;
```
A tiny optimisation: raft->max_vote is >= previous_max - 1 anyway.

sergepetrenko replied:
`raft->max_vote--` instead of `raft->max_vote = 0`.
If you were max and did `--v->count` it means either max_vote is still the same or it is decremented.

Gerold103 replied:
But we still need to do the fullscan, because we don't know if it changed or not. Anyway, updated to your version.

sergepetrenko replied:
Yeah, we do, but we'll perform less assignments.

There's another one: you may break the loop as soon as max_vote reaches old_max (basically as soon as max_vote is increased).
But this is over the top, I guess. Will make the code harder to follow. Current version looks fine.

sergepetrenko commented on src/lib/raft/raft.c:
```diff
@@ -158,6 +159,32 @@ raft_add_vote(struct raft *raft, int src, int dst)
+ * to any other instance yet.
+ */
+static void
+raft_del_vote(struct raft *raft, int src, int dst)
```
It seems we are only deleting votes from `raft->self`. Let's remove `src` from parameters then?

Gerold103 replied:
Done.

sergepetrenko commented on src/lib/raft/raft.c:
```diff
@@ -695,14 +721,51 @@ raft_worker_handle_io(struct raft *raft)
+		 * can be <= candidate's now, but that can change after the WAL
+		 * queue is flushed.
+		 */
+		if (raft->volatile_vote == raft->vote ||
```
I guess `volatile_vote == vote` is impossible in this branch (the only case is volatile_vote == 0).

Gerold103 replied:
It is possible if you vote for self + bump term in the same message. Anyway, I reworked this part.

sergepetrenko commented on src/lib/raft/raft.c:
```diff
@@ -695,14 +721,51 @@ raft_worker_handle_io(struct raft *raft)
+		 * When vote for self, it is always acceptable. Can persist it
+		 * together with the term.
+		 */
+		if (raft->volatile_vote == raft->self)
```
Maybe unify voting for self and for someone else?
Otherwise raft behaves differently based on some specifics not related to raft.
I think we could always persist term and vote separately

Gerold103 replied:
This makes the patch notably more complicated mostly due to much bigger diff and because users start seeing intermediate states on the current node via `on_election` callbacks. Also I still need to check for self here to avoid calling `raft_can_vote_for()` on own old vclock. Otherwise can fail a check for voting for self and the log about looks really weird. Done anyway, see the new version.

sergepetrenko commented on test/replication-luatest/gh_7253_election_long_wal_write_test.lua:
```diff
@@ -171,3 +171,86 @@ g.test_fence_during_confirm_wal_write = function(g)
+    -- Server1 gets the new term via server2.
+    --
+    g.server1:wait_election_term(term + 1)
+    -- Restore server3 original params.
```
Maybe check `box.info.election` on `server2` as well? Otherwise looks like server3 could still win and pass the test by chance

Gerold103 replied:
Done.

sergepetrenko reviewed (changes requested):
Hi! Thanks for the patch!
Please find my comments below.

Gerold103 commented on src/box/box.cc:
```diff
@@ -1806,40 +1806,39 @@ box_wait_quorum(uint32_t lead_id, int64_t target_lsn, int quorum,
  */
 static int
-box_run_elections(void)
+box_promote_try_until_cant(void)
```
I was hoping you wouldn't notice, damn. Done in the new version.

Gerold103 commented on src/box/raft.c:
```diff
@@ -451,34 +456,38 @@ box_raft_wait_term_outcome_f(struct trigger *trig, void *event)
 
 int
-box_raft_wait_term_outcome(void)
+box_raft_promote_try(void)
```
Done.

Gerold103 commented on src/box/raft.c:
```diff
@@ -451,34 +456,38 @@ box_raft_wait_term_outcome_f(struct trigger *trig, void *event)
+	trigger_create(&trig, box_raft_promote_try_f, &ctx, NULL);
 	raft_on_update(raft, &trig);
-	/*
-	 * XXX: it is not a good idea not to have a timeout here. If all nodes
```
Maybe we can. I didn't care about that here though. Sending nacks might be complicated though. It would mean a response to a specific notification, which means they would need to have some kind of request IDs. Don't see an obvious solution here.

grafin commented on test/replication-luatest/gh_7253_election_long_wal_write_test.lua:
```diff
@@ -148,8 +165,97 @@ g.test_fence_during_confirm_wal_write = function(g)
+    g.server3:exec(function()
+        local replication = table.copy(box.cfg.replication)
+        rawset(_G, 'old_replication', table.copy(replication))
+        table.remove(replication, 1)
```
Why `table.remove here` and `replication[3] = nil` several lines before?

grafin commented on test/replication-luatest/gh_7253_election_long_wal_write_test.lua:
```diff
@@ -0,0 +1,155 @@
+g.test_fence_during_confirm_wal_write = function(g)
+    --
+    -- Make the following topology:
+    --   server2 <-> server1 <- server3
```
Looks like this comment is from previous version, with 3 servers, instead of 2.

grafin commented on test/replication-luatest/gh_7253_election_long_wal_write_test.lua:
```diff
@@ -0,0 +1,155 @@
+    g.server2:exec(function()
+        local replication = table.copy(box.cfg.replication)
+        rawset(_G, 'old_replication', table.copy(replication))
+        replication[3] = nil
```
With 2 servers replication[3] is nil already, this whole exec call looks redundant.

grafin commented on test/replication-luatest/gh_7253_election_long_wal_write_test.lua:
```diff
@@ -0,0 +1,155 @@
+    server_unblock_wal_write(g.server2)
+    server_wait_wal_is_blocked(g.server1)
+    --
+    -- Server3 sends a new term to server1.
```
Same as above.

grafin reviewed (changes requested):
Thanks for the patchset!

grafin reviewed (approved):
LGTM.

sergepetrenko commented on test/replication-luatest/gh_7253_election_long_wal_write_test.lua:
```diff
@@ -0,0 +1,155 @@
+    server_unblock_wal_write(g.server2)
+    server_wait_wal_is_blocked(g.server1)
+    --
+    -- Server3 sends a new term to server1.
```
It's server2, not server 3

Gerold103 replied:
Indeed, thanks for noticing, fixed.
<details><summary>diff</summary>

```Diff
diff --git a/test/replication-luatest/gh_7253_election_long_wal_write_test.lua b/test/replication-luatest/gh_7253_election_long_wal_write_test.lua
index 3680f0cea..6ea5208eb 100644
--- a/test/replication-luatest/gh_7253_election_long_wal_write_test.lua
+++ b/test/replication-luatest/gh_7253_election_long_wal_write_test.lua
@@ -84,16 +84,6 @@ g.after_all(function(g)
 end)
 
 g.test_fence_during_confirm_wal_write = function(g)
-    --
-    -- Make the following topology:
-    --   server2 <-> server1 <- server3
-    --
-    g.server2:exec(function()
-        local replication = table.copy(box.cfg.replication)
-        rawset(_G, 'old_replication', table.copy(replication))
-        replication[3] = nil
-        box.cfg{replication = replication}
-    end)
     --
     -- Server1 starts a txn.
     --
@@ -113,7 +103,7 @@ g.test_fence_during_confirm_wal_write = function(g)
     server_unblock_wal_write(g.server2)
     server_wait_wal_is_blocked(g.server1)
     --
-    -- Server3 sends a new term to server1.
+    -- Server2 sends a new term to server1.
     --
     local term = g.server1:election_term()
     fiber.create(g.server2.exec, g.server2, function()
@@ -142,11 +132,9 @@ g.test_fence_during_confirm_wal_write = function(g)
     --
     g.server2:exec(function()
         box.cfg{
-            replication = _G.old_replication,
-            election_timeout = box.NULL,
             election_mode = 'voter',
+            election_timeout = box.NULL,
         }
-        _G.old_replication = nil
     end)
     g.server1:exec(function()
         box.ctl.promote()

```
</details>

sergepetrenko commented on test/replication-luatest/gh_7253_election_long_wal_write_test.lua:
```diff
@@ -0,0 +1,155 @@
+    g.server2:exec(function()
+        local replication = table.copy(box.cfg.replication)
+        rawset(_G, 'old_replication', table.copy(replication))
+        replication[3] = nil
```
Sorry for nitpicking, but you don't have server3 in 3rd commit yet, so it should appear in comments and in replication[3] only in next commit.

Gerold103 replied:
True, fixed in the diff in the comment above.

sergepetrenko reviewed (commented):
I haven't checked the last commit yet, will return to it tomorrow.

Gerold103 commented:
CI is red because the fresh checkpatch got a bug: https://github.com/tarantool/checkpatch/issues/37.
Also it seems github is broken. I can see @grafin comments in my mail, but not on this webpage. Answering them here:

> Why table.remove here and replication[3] = nil several lines before?

`tbl[N] = nil` replaces the item with `nil`. When done on tail, it works as 'pop last'. When done on a non-last item, it leaves a hole in the table. But `box.cfg.replication` should be a proper array without holes. Anyway, now I updated both cases to use `table.remove`, wouldn't hurt.

> * Looks like this comment is from previous version, with 3 servers, instead of 2.
> * With 2 servers replication[3] is nil already, this whole exec call looks redundant.
> * Same as above.

Thanks for noticing, fixed.
<details><summary>diff</summary>

```Diff
diff --git a/test/replication-luatest/gh_7253_election_long_wal_write_test.lua b/test/replication-luatest/gh_7253_election_long_wal_write_test.lua
index 3680f0cea..6ea5208eb 100644
--- a/test/replication-luatest/gh_7253_election_long_wal_write_test.lua
+++ b/test/replication-luatest/gh_7253_election_long_wal_write_test.lua
@@ -84,16 +84,6 @@ g.after_all(function(g)
 end)
 
 g.test_fence_during_confirm_wal_write = function(g)
-    --
-    -- Make the following topology:
-    --   server2 <-> server1 <- server3
-    --
-    g.server2:exec(function()
-        local replication = table.copy(box.cfg.replication)
-        rawset(_G, 'old_replication', table.copy(replication))
-        replication[3] = nil
-        box.cfg{replication = replication}
-    end)
     --
     -- Server1 starts a txn.
     --
@@ -113,7 +103,7 @@ g.test_fence_during_confirm_wal_write = function(g)
     server_unblock_wal_write(g.server2)
     server_wait_wal_is_blocked(g.server1)
     --
-    -- Server3 sends a new term to server1.
+    -- Server2 sends a new term to server1.
     --
     local term = g.server1:election_term()
     fiber.create(g.server2.exec, g.server2, function()
@@ -142,11 +132,9 @@ g.test_fence_during_confirm_wal_write = function(g)
     --
     g.server2:exec(function()
         box.cfg{
-            replication = _G.old_replication,
-            election_timeout = box.NULL,
             election_mode = 'voter',
+            election_timeout = box.NULL,
         }
-        _G.old_replication = nil
     end)
     g.server1:exec(function()
         box.ctl.promote()
```
</details>

sergepetrenko commented on src/lib/raft/raft.c:
```diff
@@ -695,14 +725,45 @@ raft_worker_handle_io(struct raft *raft)
+		 * Skip self. When vote was issued, own vclock could be smaller,
+		 * but that doesn't matter. Can always vote for self. Not having
+		 * this special case still works, but the node might log that
+		 * it canceled a vote for self, which is confusing.
```
> Not having this special case still works, but the node might log that it canceled a vote for self, which is confusing.

I'm not sure how it's supposed to work without that special case. Seems `raft.vote` is zero, and `raft->votalile_vote` is `self`. This shouldn't make the node become candidate. Besides, the node might forget that it voted for self after a restart.

Let's just drop that comment part.

Gerold103 replied:
> This shouldn't make the node become candidate

The node is a candidate if it is configured as a candidate. In `end_dump:` it will see that vote is zero, no leader, the node is a candidate, and it simply retries voting for self. I discovered it when tried to drop the special case to see what happens.

I would rather keep the comment, otherwise in the future we won't remember all the reasons why this check is needed. I tried to reword it:
```Diff
@@ -747,8 +747,9 @@ end_dump:
 		/*
 		 * Skip self. When vote was issued, own vclock could be smaller,
 		 * but that doesn't matter. Can always vote for self. Not having
-		 * this special case still works, but the node might log that
-		 * it canceled a vote for self, which is confusing.
+		 * this special case still works if the node is configured as a
+		 * candidate, but the node might log that it canceled a vote for
+		 * self, which is confusing.
 		 */
 		if (raft->volatile_vote == raft->self)
 			goto do_dump_with_vote;
```

sergepetrenko replied:
> The node is a candidate if it is configured as a candidate. In end_dump: it will see that vote is zero, no leader, the node is a candidate, and it simply retries voting for self. I discovered it when tried to drop the special case to see what happens.

I see now, thanks for the explanation!

sergepetrenko commented on test/replication-luatest/gh_7253_election_long_wal_write_test.lua:
```diff
@@ -136,8 +153,97 @@ g.test_fence_during_confirm_wal_write = function(g)
+    -- Server2 gets a foreign txn going to WAL too long.
+    --
+    server_block_next_wal_write(g.server2)
+    local f = fiber.new(g.server1.exec, g.server1, function()
```
I remember there were issues with hanging one `exec` and trying to perform another one to the same server later.
I'd rather make the exec return immediately by doing something like
```lua
g.server1:exec(function()
    require('fiber').new(....)
end)
```
But if the current version works fine for you, then never mind. Just have to be careful with that.

Gerold103 replied:
AFAIK the hang can only happen if:
- The first exec killed the server;
- The first exec made the server run a busy-loop without yields;
- The first exec caused connection termination and the server rebound its server to another port;
- The server was configured to have only one fiber for serving iproto requests and it is still busy with the first request.

In this case neither of these should happen. This hack about returning fiber ids I don't like, because it forces to do `require('fiber')` on the server side every time when you want to start some async work there. It just pads the code out. I don't mind seeing it in other patches, but would want to avoid in mine.

sergepetrenko replied:
> In this case neither of these should happen. 

Yes, I was thinking the same, but still I remember hitting this issue in either one of mine or @grafin's patches. I don't remember which one.

Anyway, leave this version if you like it. I see that it works now, so I don't mind.

sergepetrenko reviewed (commented):
Thanks for the fixes! Just a couple of nits remaining.

sergepetrenko reviewed (approved):
Thanks for the fixes!
LGTM.
